components/raftstore/src/store/fsm/peer.rs

```diff
@@ -753,157 +743,170 @@ where
-            ccs.set_node_id(peer_id);
-            cc.set_transition(eraftpb::ConfChangeTransition::Auto);
-            cc.mut_changes().push(ccs);
-            if let Err(e) = self.fsm.peer.raft_group.apply_conf_change(&cc) {
-                panic!("fail to apply conf change for unsafe recover {:?}", e);
+    }
+
+    fn unsafe_recovery_demote_failed_voters(
+        &mut self,
+        syncer: UnsafeRecoveryExecutePlanSyncer,
+        failed_voters: Vec<metapb::Peer>,
+    ) {
+        if let Some(req) =
+            demote_failed_voters_request(self.region(), &self.fsm.peer.peer, failed_voters)
+        {
+            info!(
+                "Unsafe recovery, demoting failed voters";
+                "region_id" => self.region().get_id(),
+                "peer_id" => self.fsm.peer.peer.get_id(),
+                "req" => ?req);
+            let failed = Arc::new(Mutex::new(false));
+            let failed_clone = failed.clone();
+            let callback = Callback::<EK::Snapshot>::write(Box::new(move |resp| {
+                if resp.response.get_header().has_error() {
+                    *failed_clone.lock().unwrap() = true;
+                    error!(
+                        "Unsafe recovery, fail to finish demotion";
+                        "err" => ?resp.response.get_header().get_error(),
+                    );
+                }
+            }));
+            self.propose_raft_command_internal(req, callback, DiskFullOpt::AllowedOnAlmostFull);
+            if !*failed.lock().unwrap() {
+                self.fsm.peer.unsafe_recovery_state =
+                    Some(UnsafeRecoveryState::DemoteFailedVoters {
+                        syncer,
+                        failed_voters: vec![], // No longer needed since here.
+                        target_index: self.fsm.peer.raft_group.raft.raft_log.last_index(),
+                        demote_after_exit: false,
+                    });
             }
         }
-        self.fsm
-            .peer
-            .peer_heartbeats
-            .retain(|&k, _| new_peer_list.contains(&k));
-        self.fsm
-            .peer
-            .peers_start_pending_time
-            .retain(|&(k, _)| new_peer_list.contains(&k));
-        for peer in to_be_removed {
-            self.fsm.peer.remove_peer_from_cache(peer);
+    }
+
+    fn on_unsafe_recovery_destroy(&mut self, syncer: UnsafeRecoveryExecutePlanSyncer) {
+        if self.fsm.peer.unsafe_recovery_state.is_some() {
+            warn!(
+                "Unsafe recovery, can't destroy, another plan is executing in progress";
+                "region_id" => self.region_id(),
+                "peer_id" => self.fsm.peer_id(),
+            );
+            syncer.abort();
+            return;
         }
-        self.fsm.peer.post_split();
-        self.fsm.reset_hibernate_state(GroupState::Chaos);
-        self.register_raft_base_tick();
+        self.fsm.peer.unsafe_recovery_state = Some(UnsafeRecoveryState::Destroy(syncer));
```

Comment:
Should we check peer role here because this command should not be sent to a force leader as I understand?

Reply:
It is technically possible to destroy force leader peer, since unsafe recovery is a multi-rounds process, after the region applies to the newest index, we may find it conflict with another existing region, thus issue a destroy.